tensorflow/contrib/s3/s3_file_system.cc

```diff
@@ -0,0 +1,578 @@
+  string bucket, object;
+  TF_RETURN_IF_ERROR(ParseS3Path(dirname, true, &bucket, &object));
+
+  if (object.empty()) {
+    Aws::S3::S3Client s3Client(GetDefaultClientConfig());
+    Aws::S3::Model::HeadBucketRequest headBucketRequest;
+    headBucketRequest.WithBucket(bucket.c_str());
+    auto headBucketOutcome = s3Client.HeadBucket(headBucketRequest);
+    if (!headBucketOutcome.IsSuccess()) {
+      return errors::NotFound("The bucket ", bucket, " was not found.");
+    }
+    return Status::OK();
+  }
+  string filename = dirname;
+  if (filename.back() != '/') {
+    filename.push_back('/');
+  }
+  std::unique_ptr<WritableFile> file;
+  TF_RETURN_IF_ERROR(NewWritableFile(filename, &file));
+  TF_RETURN_IF_ERROR(file->Close());
+  return Status::OK();
+}
+
+Status S3FileSystem::DeleteDir(const string& dirname) {
+  string bucket, object;
+  TF_RETURN_IF_ERROR(ParseS3Path(dirname, false, &bucket, &object));
+
+  Aws::S3::S3Client s3Client(GetDefaultClientConfig());
+  string prefix = object;
+  if (prefix.back() != '/') {
+    prefix.push_back('/');
+  }
+  Aws::S3::Model::ListObjectsRequest listObjectsRequest;
+  listObjectsRequest.WithBucket(bucket.c_str())
+      .WithPrefix(prefix.c_str())
+      .WithMaxKeys(2);
+  listObjectsRequest.SetResponseStreamFactory(
+      []() { return Aws::New<Aws::StringStream>(S3FileSystemAllocationTag); });
+  auto listObjectsOutcome = s3Client.ListObjects(listObjectsRequest);
+  if (listObjectsOutcome.IsSuccess()) {
+    auto contents = listObjectsOutcome.GetResult().GetContents();
+    if (contents.size() > 1 ||
+        (contents.size() == 1 && contents[0].GetKey() != prefix.c_str())) {
+      return errors::FailedPrecondition("Cannot delete a non-empty directory.");
+    }
+    if (contents.size() == 1 && contents[0].GetKey() == prefix.c_str()) {
+      string filename = dirname;
+      if (filename.back() != '/') {
+        filename.push_back('/');
+      }
+      return DeleteFile(filename);
+    }
+  }
+  return Status::OK();
+}
+
+Status S3FileSystem::GetFileSize(const string& fname, uint64* file_size) {
+  FileStatistics stats;
+  TF_RETURN_IF_ERROR(this->Stat(fname, &stats));
+  *file_size = stats.length;
+  return Status::OK();
+}
+
+Status S3FileSystem::RenameFile(const string& src, const string& target) {
+  string src_bucket, src_object, target_bucket, target_object;
+  TF_RETURN_IF_ERROR(ParseS3Path(src, false, &src_bucket, &src_object));
+  TF_RETURN_IF_ERROR(
+      ParseS3Path(target, false, &target_bucket, &target_object));
+  if (src_object.back() == '/') {
+    if (target_object.back() != '/') {
+      target_object.push_back('/');
+    }
+  } else {
+    if (target_object.back() == '/') {
+      target_object.pop_back();
+    }
+  }
+
+  Aws::S3::S3Client s3Client(GetDefaultClientConfig());
+
+  Aws::S3::Model::CopyObjectRequest copyObjectRequest;
+  Aws::S3::Model::DeleteObjectRequest deleteObjectRequest;
+
+  Aws::S3::Model::ListObjectsRequest listObjectsRequest;
```

Comment:
Thanks @jhseu  In the GCS file systems the RenameFile also does the directory rename:
https://github.com/tensorflow/tensorflow/blob/0440c77935915fce0d9286ce5e7407b31c6f4975/tensorflow/core/platform/cloud/gcs_file_system.cc#L1154-L1168